mk/spksrc.spk.mk

```diff
@@ -14,8 +14,14 @@ TC = syno$(ARCH_SUFFIX)
 else
 SPK_ARCH = noarch
 SPK_NAME_ARCH = noarch
+ifeq ($(shell expr "$(TCVERSION)" \>= 7.0),1)
```

Comment:
`expr` doesn't really support float.
Another method could be using the `sort -V` for versioning like I did for the kernel side of things:
```
# olddefconfig is not available <= 3.2
ifeq ($(shell printf '%s\n' "$(TC_KERNEL)" "3.3" | sort -V | head -1),$(TC_KERNEL))
        @$(MSG) "oldconfig OLD style... $(TC_KERNEL) <= 3.2"
        $(RUN) yes "" | $(MAKE) oldconfig
else
        $(RUN) $(MAKE) olddefconfig
endif
```

Reply:
Hm, It works as it should and is easy to read. If I look at your version I have to take longer to understand it and uses 3 tools.

I've read the man page and skimmed `info expr` I don't see anywhere where floats are unsupported... where 
 
If we need to change can I propose `bc` instead?

```sh
docker run -it --rm --name spksrc -v $HOME/git/spksrc:/spksrc synocommunity/spksrc
root@04c82828bdd1:/spksrc# expr "6.9" \>= 7.0
0
root@04c82828bdd1:/spksrc# expr "7.1" \>= 7.0
1

root@04c82828bdd1:/spksrc# echo "6.9 >= 7.0" | bc -l
0
root@04c82828bdd1:/spksrc# echo "7.1 >= 7.0" | bc -l
1

root@04c82828bdd1:/spksrc# expr "7.1234" \> 7.123
1
root@04c82828bdd1:/spksrc# expr "7.0234" \> 7.123
0
```

Reply:
Interesting a snipit from `info expr`
```
‘< <= = == != >= >’
     Compare the arguments and return 1 if the relation is true, 0
     otherwise.  ‘==’ is a synonym for ‘=’.  ‘expr’ first tries to
     convert both arguments to integers and do a numeric comparison; if
     either conversion fails, it does a lexicographic comparison using
     the character collating sequence specified by the ‘LC_COLLATE’
     locale.
```

Reply:
So where would a lexicographic comparison fail as apposed to a float? If anything it eliminates float rounding errors.

Reply:
Ok I found a rare case:

$ expr 7 \>= 7
1

$ expr 7  \>= 7.0
0

Reply:
I hit this rather interesting issue when comparing kernel version 3.10 with 3.2 ...

Reply:
Ah ok the problem isn't with floats (mathematically 3.10 is less than 3.2, float or no float will help). Maybe we can wrap your code into a function? Will that work in make files? e.g. `version_lt 3.10 3.2`?  https://stackoverflow.com/questions/16989598/comparing-php-version-numbers-using-bash/24067243

Reply:
Perhaps something as simple as (totally untested):
```
version_comparison:
ifeq ($(shell printf '%s\n' "$(VERSION)" "$(VERSION_TARGET)" | sort -V | head -1),$(VERSION_TARGET))
        echo 1
endif

...

VERSION=a.b
VERSION_TARGET=b.c

ifeq ($(strip $(version_comparison)),1)
        @$(MSG) "$(VERSION) is >= to $(VERSION_TARGET)"
else
        @$(MSG) "$(VERSION) is <= to $(VERSION_TARGET)"
fi
```

Reply:
With the help of https://stackoverflow.com/a/60021707
```MAKE
# version_lt = $(shell if printf '%s\n%s\n' "$(1)" "$(2)" | sort --version-sort --check=quiet ; then echo 1; fi)
# version_gt = $(shell if printf '%s\n%s\n' "$(2)" "$(1)" | sort --version-sort --check=quiet ; then echo 1; fi)
version_le = $(shell if printf '%s\n' "$(1)" "$(2)" | sort -V -C ; then echo 1; fi)
version_ge = $(shell if printf '%s\n' "$(2)" "$(1)" | sort -V -C ; then echo 1; fi)

$(info *********)
VERSION=7.0
ifeq ($(call version_le, ${VERSION}, 6.0),1)
$(info "$(VERSION) is <= 6.0")
else
$(info "$(VERSION) is > 6.0")
endif
$(info *********)

ifeq ($(call version_ge, ${VERSION}, 6.0),1)
$(info "$(VERSION) is >= 6.0")
else
$(info "$(VERSION) is < 6.0")
endif
$(info *********)
```

Results
```
VERSION=3.10
*********
"3.10 is > 3.2"
*********
"3.10 is >= 3.2"
*********


VERSION=4.0
*********
"4.0 is <= 6.0"
*********
"4.0 is < 6.0"
*********

VERSION=6.0
*********
"6.0 is <= 6.0"
*********
"6.0 is >= 6.0"
*********
```

Reply:
Hm, which of these easier to read?

```make
ifneq ($(call version_ge, ${TCVERSION}, 7.0),1)
# old code
endif
ifeq ($(call version_le, ${TCVERSION}, 6.2.3),1)
# old code
endif
```

Reply:
maybe we need an additional function `version_lt` (less than) to implement like this
``` makefile
ifeq ($(call version_lt, ${TCVERSION}, 7.0),1)
# old code
endif
```

This will evaluate as expected, even when version 6.2.4 or other would be released.
For this reason I would prefer the first solution of yours.
_And it is better than the use of any magic version like 6.99999)_

Reply:
@hgy59 Thanks that was my concern as well, that is why I've already used first version. I just noticed there is a lot of "if not equal to" in the Makefiles. Not a big issue but I have been caught out on it on occasion, It's easy to miss the little `n`. 
From your feedback I've added the `version_lt` function as well.